File: pi42/user_data.py
"""
User Data API endpoints for Pi42.
"""

from typing import Dict, Any, Optional, List


class UserDataAPI:
    """
    User Data API client for accessing Pi42 user-specific data.
    """

    def __init__(self, client):
        """
        Initialize the User Data API client.

        Args:
            client: The main Pi42Client instance
        """
        self._client = client

    def get_trade_history(self, start_timestamp: Optional[int] = None,
                         end_timestamp: Optional[int] = None,
                         sort_order: Optional[str] = None,
                         page_size: Optional[int] = None,
                         symbol: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Retrieve the trade history for a user.

        Args:
            start_timestamp: Start timestamp for filtering trade history
            end_timestamp: End timestamp for filtering trade history
            sort_order: Sorting order ('asc' or 'desc')
            page_size: Number of results to return per page
            symbol: Trading pair symbol to filter trade history

        Returns:
            List[Dict]: List of trade history items
        """
        endpoint = "/v1/user-data/trade-history"
        
        params = {}
        
        if start_timestamp:
            params["startTimestamp"] = start_timestamp
        if end_timestamp:
            params["endTimestamp"] = end_timestamp
        if sort_order:
            params["sortOrder"] = sort_order
        if page_size:
            params["pageSize"] = page_size
        if symbol:
            params["symbol"] = symbol
            
        return self._client.get_request(endpoint, params)

    def get_transaction_history(self, start_timestamp: Optional[int] = None,
                               end_timestamp: Optional[int] = None,
                               sort_order: Optional[str] = None,
                               page_size: Optional[int] = None,
                               symbol: Optional[str] = None,
                               trade_id: Optional[int] = None,
                               position_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Retrieve the transaction history for a user.

        Args:
            start_timestamp: Start timestamp for filtering transaction history
            end_timestamp: End timestamp for filtering transaction history
            sort_order: Sorting order ('asc' or 'desc')
            page_size: Number of records to return per page
            symbol: Trading symbol to filter the transaction history
            trade_id: Specific trade ID to filter the transaction history
            position_id: Specific position ID to filter the transaction history

        Returns:
            List[Dict]: List of transaction history items
        """
        endpoint = "/v1/user-data/transaction-history"
        
        params = {}
        
        if start_timestamp:
            params["startTimestamp"] = start_timestamp
        if end_timestamp:
            params["endTimestamp"] = end_timestamp
        if sort_order:
            params["sortOrder"] = sort_order
        if page_size:
            params["pageSize"] = page_size
        if symbol:
            params["symbol"] = symbol
        if trade_id:
            params["tradeId"] = trade_id
        if position_id:
            params["positionId"] = position_id
            
        return self._client.get_request(endpoint, params)

    def create_listen_key(self) -> Dict[str, str]:
        """
        Create a new listen key for WebSocket connections.

        Returns:
            Dict: Response containing the listen key
        """
        endpoint = "/v1/retail/listen-key"
        
        return self._client.post_request(endpoint)

    def update_listen_key(self) -> str:
        """
        Update the listen key for WebSocket connections.

        Returns:
            str: Confirmation message
        """
        endpoint = "/v1/retail/listen-key"
        
        return self._client.put_request(endpoint)

    def delete_listen_key(self) -> str:
        """
        Delete the listen key for WebSocket connections.

        Returns:
            str: Confirmation message
        """
        endpoint = "/v1/retail/listen-key"
        
        return self._client.delete_request(endpoint)

File: user_data.go
package pi42

import (
	"encoding/json"
	"fmt"
	"strconv"
)

// UserDataAPI provides access to user-specific data endpoints
type UserDataAPI struct {
	client *Client
}

// NewUserDataAPI creates a new User Data API instance
func NewUserDataAPI(client *Client) *UserDataAPI {
	return &UserDataAPI{client: client}
}

// DataQueryParams represents common parameters for data queries
type DataQueryParams struct {
	StartTimestamp int64  `json:"startTimestamp,omitempty"`
	EndTimestamp   int64  `json:"endTimestamp,omitempty"`
	SortOrder      string `json:"sortOrder,omitempty"`
	PageSize       int    `json:"pageSize,omitempty"`
	Symbol         string `json:"symbol,omitempty"`
}

// GetTradeHistory retrieves the trade history for a user
func (api *UserDataAPI) GetTradeHistory(params DataQueryParams) ([]map[string]interface{}, error) {
	endpoint := "/v1/user-data/trade-history"

	queryParams := make(map[string]string)

	if params.StartTimestamp > 0 {
		queryParams["startTimestamp"] = strconv.FormatInt(params.StartTimestamp, 10)
	}
	if params.EndTimestamp > 0 {
		queryParams["endTimestamp"] = strconv.FormatInt(params.EndTimestamp, 10)
	}
	if params.SortOrder != "" {
		queryParams["sortOrder"] = params.SortOrder
	}
	if params.PageSize > 0 {
		queryParams["pageSize"] = strconv.Itoa(params.PageSize)
	}
	if params.Symbol != "" {
		queryParams["symbol"] = params.Symbol
	}

	data, err := api.client.Get(endpoint, queryParams, false)
	if err != nil {
		return nil, err
	}

	var result []map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// TransactionHistoryParams extends DataQueryParams with additional fields
type TransactionHistoryParams struct {
	DataQueryParams
	TradeID    int    `json:"tradeId,omitempty"`
	PositionID string `json:"positionId,omitempty"`
}

// GetTransactionHistory retrieves the transaction history for a user
func (api *UserDataAPI) GetTransactionHistory(params TransactionHistoryParams) ([]map[string]interface{}, error) {
	endpoint := "/v1/user-data/transaction-history"

	queryParams := make(map[string]string)

	if params.StartTimestamp > 0 {
		queryParams["startTimestamp"] = strconv.FormatInt(params.StartTimestamp, 10)
	}
	if params.EndTimestamp > 0 {
		queryParams["endTimestamp"] = strconv.FormatInt(params.EndTimestamp, 10)
	}
	if params.SortOrder != "" {
		queryParams["sortOrder"] = params.SortOrder
	}
	if params.PageSize > 0 {
		queryParams["pageSize"] = strconv.Itoa(params.PageSize)
	}
	if params.Symbol != "" {
		queryParams["symbol"] = params.Symbol
	}
	if params.TradeID > 0 {
		queryParams["tradeId"] = strconv.Itoa(params.TradeID)
	}
	if params.PositionID != "" {
		queryParams["positionId"] = params.PositionID
	}

	data, err := api.client.Get(endpoint, queryParams, false)
	if err != nil {
		return nil, err
	}

	var result []map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// CreateListenKey creates a new listen key for Socketio connections
func (api *UserDataAPI) CreateListenKey() (map[string]string, error) {
	endpoint := "/v1/retail/listen-key"

	data, err := api.client.Post(endpoint, map[string]interface{}{}, false)
	if err != nil {
		return nil, err
	}

	var result map[string]string
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// UpdateListenKey updates the listen key for Socketio connections
func (api *UserDataAPI) UpdateListenKey() (string, error) {
	endpoint := "/v1/retail/listen-key"

	data, err := api.client.Put(endpoint, map[string]interface{}{})
	if err != nil {
		return "", err
	}

	return string(data), nil
}

// DeleteListenKey deletes the listen key for Socketio connections
func (api *UserDataAPI) DeleteListenKey() (string, error) {
	endpoint := "/v1/retail/listen-key"

	data, err := api.client.Delete(endpoint, map[string]interface{}{})
	if err != nil {
		return "", err
	}

	return string(data), nil
}
